File: backend/app/schemas/forecast.py
from datetime import date, datetime
from typing import List, Optional
from pydantic import BaseModel, Field, model_validator

class ForecastGenerateRequest(BaseModel):
    forecast_period: str = Field(..., pattern="^(7d|30d|90d|custom)$")
    date_from: Optional[date] = None
    date_to: Optional[date] = None
    @model_validator(mode="after")
    def validate_custom_range(self):
        if self.forecast_period == "custom":
            if not self.date_from or not self.date_to:
                raise ValueError("date_from and date_to are required for a custom forecast")
            if self.date_to < self.date_from:
                raise ValueError("date_to cannot be before date_from")
        return self

class ForecastProductOut(BaseModel):
    id: int; product_id: int; product_name: str; category_id: int; category_name: str
    brand: Optional[str] = None; current_stock: int; reorder_level: int; historical_sales: int
    predicted_demand: int; forecast_period: str; period_start: date; period_end: date
    confidence_score: float; accuracy: float; growth_percentage: float; recommendation: str; generated_at: datetime

class ForecastCategoryOut(BaseModel):
    category_id: int; category_name: str; total_historical_sales: int; predicted_demand: int; expected_growth_percentage: float

class ForecastDashboardOut(BaseModel):
    kpis: dict; products: List[ForecastProductOut]; categories: List[ForecastCategoryOut]
    historical_vs_forecast: List[dict]; product_demand_trend: List[dict]; category_demand_trend: List[dict]
    top_predicted_products: List[dict]; seasonal_sales_pattern: List[dict]

File: frontend/src/types/forecast.ts
export type ForecastPeriod =
    | "7d"
    | "30d"
    | "90d"
    | "custom";

export interface ForecastProduct {
    id: number;
    product_id: number;
    product_name: string;
    category_id: number;
    category_name: string;
    brand?: string;
    current_stock: number;
    reorder_level: number;
    historical_sales: number;
    predicted_demand: number;
    forecast_period: ForecastPeriod;
    period_start: string;
    period_end: string;
    confidence_score: number;
    accuracy: number;
    growth_percentage: number;
    recommendation: string;
    generated_at: string;
}

export interface ForecastCategory {
    category_id: number;
    category_name: string;
    total_historical_sales: number;
    predicted_demand: number;
    expected_growth_percentage: number;
}

export interface ForecastDashboard {
    kpis: {
        total_predicted_demand: number;
        products_expected_to_run_out: number;
        high_growth_products: number;
        slow_moving_products: number;
        forecast_accuracy: number;
    };

    products: ForecastProduct[];

    categories: ForecastCategory[];

    historical_vs_forecast: {
        product_name: string;
        historical_sales: number;
        predicted_demand: number;
    }[];

    product_demand_trend: {
        product_name: string;
        historical_sales: number;
        predicted_demand: number;
    }[];

    category_demand_trend: ForecastCategory[];

    top_predicted_products: {
        product_name: string;
        predicted_demand: number;
    }[];

    seasonal_sales_pattern: {
        month: string;
        sales: number;
    }[];
}

export interface ForecastRequest {
    forecast_period: ForecastPeriod;
    date_from?: string;
    date_to?: string;
}